### DISCUSSION TITLE
Google OAth integration with Fastapi

### DISCUSSION THREAD

**Discussion by matejko94** [fastapi-users/fastapi-users#1183] [General]
We use fastapi-users to integrate [Google auth](https://fastapi-users.github.io/fastapi-users/10.4/). By default there are defined to endpoints `/auth/google/authorize` and `/auth/google/callback`. 

So frontend which is in Angular calls `/auth/google/authorize` and as result we got URL in body. Than if we call this URL we got CORS error. After that we are not sure if this is right approach or on which order should we call API. 

Also API `/auth/google/callback` has code, code_verifier, state and error fields where we get this fields?

[![enter image description here][1]][1]

We use fastapi-users package in our backend. We have major problems with integrating Google OAuth.
URL: [google auth](https://fastapi-users.github.io/fastapi-users/10.4/)
By default there are two endpoints `/auth/google/authorize` and `/auth/google/callback`. 

What is actually a flow for these endpoints? When we call the /auth/google/authorize endpoint, the response is:

[![enter image description here][2]][2]

We tried then calling this URL again in the frontend, because if we copy paste this URL in google, the response is token and token type.
If we call this endpoint in our frontend, we get CORS error. 

Is this even right approach? What frontend should do with these two endpoints? And what to do with the /auth/google/callback endpoint?

Thanks to everyone for help

  [1]: https://i.stack.imgur.com/DyVl8.png
  [2]: https://i.stack.imgur.com/ehVMZ.png

@frankie567  Please could you help. Thank you

**Comment by frankie567** [comment]
This is your frontend that will make the API calls to the OAuth routes.

## 1. Authorization

The first thing is to generate the right authorization URL for the OAuth Provider (here, Google). From your frontend, you call the `/authorize` endpoint. You get the following response:

```json
{
    "authorization_url": "https://www.tintagel.bt/oauth/authorize?client_id=CLIENT_ID&scopes=a+b&redirect_uri=https://www.camelot.bt/oauth/callback"
}
```

You should redirect your user to `authorization_url`, so it can authenticate on Google. In JavaScript, this can be simply done with:

```js
window.location.href = authorization_url
```

## 2. Callback

After user has authenticated on Google, they'll be redirected to your application. That's the "callback". What I usually suggest is to make a redirection to a special page **on your frontend**. The URL will look like this:

```
http://app.mydomain.com/oauth-callback?code=XXX&state=YYY
```

At that moment, from your JS app, you can catch the `code` and `state` parameter from the URL and *then* make the API call to `/oauth/callback`. If everything goes well, you'll get your JWT or session cookie and can directly redirect your user to the page you want.

```mermaid
flowchart LR
    subgraph BACKEND [Backend]
       AUTH["/oauth/authorize"]
       CB["/oauth/callback"]
    end
    subgraph FRONTEND [Frontend]
        FAUTH["/authorize"]
        FCB["/oauth-callback"]
    end
    GOOGLE[Google]
    FAUTH -- asks URL --> AUTH
    AUTH -- returns URL --> FAUTH
    FAUTH -- redirects to URL --> GOOGLE
    GOOGLE -- redirects --> FCB
    FCB -- forwards code --> CB
    CB -- returns token --> FCB
```

## 3. CORS

You may have CORS errors from your browsers if your frontend is not on the same domain as your backend. That's a normal security measure from your browser.

If you need to *relax* that security measure, you have to setup a `CORSMiddleware` to authorize your frontend to call your backend. Ref: https://fastapi.tiangolo.com/tutorial/cors/

**Reply by zigax1** [reply]
Thank you [frankie567](https://github.com/frankie567) for your response.

I am developing front end with Angular for this specific case.

When I do this:

this.boxesService.googleLogin().subscribe((res) => {
      window.location.href = res.authorization_url;
    });

I am redirected to new page, which url is: http://webapp-eture.azurewebsites.net/auth/google/callback?state=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiJmYXN0YXBpLXVzZXJzOm9hdXRoLXN0YXRlIiwiZXhwIjoxNjgwMzM2NTUxfQ.Se10d-Sp4HudeSEnwvB5aRCx4GANhqhOw6TpQm3UEng&code=4%2F0AVHEtk5M5NOKtgyWcNQ1gkxYEqLg4kVSHGZ22qoF4PWDsf8uEo0KcAHaSdu8WjCQoxWHtQ&scope=email+profile+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.profile+openid+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.email&authuser=0&prompt=none
But the domain name is backend's URL and not the frontend's url so I can not create new page in my Angular app to fetch the url parameters. I can't do anything with this url. If i use window.locaiton.href i get the token in completely separate page and url and I can't get this token from frontend.. Do you have any example for using this with Angular or are you open for shor google meet, max 10 minutes?

This would be very helpful to us and we would definetely buy you a coffe, since this is driving us crazy

**Reply by frankie567** [reply]
You'll need to set a custom redirect URI. It can be set in argument of `get_oauth_router`:

https://github.com/fastapi-users/fastapi-users/blob/master/fastapi_users/fastapi_users.py#L91-113

Don't forget to add it as allowed redirect URI on Google's side.

I just notice that this argument is not explicitly documented, this needs to be improved.

**Reply by MaslinuPoimal** [reply]
Hi @frankie567 

thanks for the very detailed explanation. I did not see it initially and had to figure it out myself, it would probably be great to add this to your OAuth documentation, this clarifies a ton.

However, I still have the following problem: when I intercept the code and state in the gauth callback and make a get request to the authorize endpoint, I get the following error in the backend:

> httpx_oauth.oauth2.GetAccessTokenError: {'error': 'redirect_uri_mismatch', 'error_description': 'Bad Request'}

I am aware this is not the exact way you described (making the same endpoint on the front-end too, redirecting to that and then posting from there), but does that make a difference? Both my front and backend are running on localhost so the redirect URI should be the same. I also did not set any additional URLs in the router which seems to be a source of problems sometimes.

**Reply by MaslinuPoimal** [reply]
NVM, I got lazy. Dropped the pre-made frontend package and just followed what is described here and it worked wonderfully. Thanks for hte great write-up, I think it should really be on the OAuth authentification page.

**Reply by vanga** [reply]
Would be nice if this flow of having frontend URI as redirect uri, passing the state and code to backend etc documented somewhere in fastapi-users documentation(I searched quite a bit and it was not obvious that this is the recommended flow). Would have saved a ton of time for me.

**Reply by sage-RJ** [reply]
Very appreciative of this discussion. Wish I found it sooner as it would have saved me time as well. [Documentation](https://fastapi-users.github.io/fastapi-users/latest/configuration/oauth/#generate-routers) goes over usage of associate_by_email and is_verified_by_default for get_oauth_router, and skips over redirect_url.
